File: vulntest/tier1-single/stack-overflow/csharp/source/Vuln.cs
// Tier 1 — stack-overflow (C# / .NET variant).
//
// `unsafe` block + `stackalloc` + raw pointer write past the end of
// the buffer. The CLR does not bounds-check raw pointer arithmetic
// inside an unsafe block — it's identical to C semantics, just
// scoped within the unsafe region.
//
// Knowledge: [[Memory/Knowledge/hw_stack_overflow_mechanics]]
// CWE-121.
//
// Build: csc Vuln.cs /unsafe
// Run:   Vuln.exe AAAAAAAA...

using System;

class Vuln {
    static unsafe void Greet(string name) {
        byte* buf = stackalloc byte[64];
        int i = 0;
        foreach (char c in name) {
            buf[i++] = (byte)c;          // sink: no bound on i
        }
    }

    static int Main(string[] args) {
        if (args.Length < 1) { Console.Error.WriteLine("usage"); return 1; }
        Greet(args[0]);
        return 0;
    }
}

File: vulntest/tier1-single/stack-overflow/c/source/vuln.c
/*
 * Tier 1 — stack buffer overflow (C variant).
 *
 * Single bug: strcpy() into a fixed 64-byte stack buffer with no
 * length check. argv[1] is the attacker-controlled source.
 *
 * Knowledge: [[Memory/Knowledge/hw_stack_overflow_mechanics]]
 * CWE-121: Stack-based Buffer Overflow.
 */
#include <stdio.h>
#include <string.h>

void greet(const char *name) {
    char buf[64];
    strcpy(buf, name);              /* sink: unbounded copy */
    printf("Hello, %s\n", buf);
}

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "usage: %s <name>\n", argv[0]);
        return 1;
    }
    greet(argv[1]);                 /* source: argv → tainted */
    return 0;
}
